system-parachains/encointer/src/migrations_fix.rs

```diff
@@ -17,173 +17,114 @@
+		use sp_std::{vec, vec::Vec};
+
+		const INVULNERABLE_A: [u8; 32] =
+			hex!("5e962096da68302d5c47fce0178d72fab503c4f00a3f1df64856748f0d9dd51e");
+		const INVULNERABLE_B: [u8; 32] =
+			hex!("0cecb8d1c2c744ca4c5cea57f5d6c40238f4dad17afa213672b8b7d43b80a659");
+		const INVULNERABLE_C: [u8; 32] =
+			hex!("ca1951a3c4e100fb5a899e7bae3ea124491930a72000c5e4b2775fea27ecf05d");
+		const INVULNERABLE_D: [u8; 32] =
+			hex!("484b443bd95068b860c92b0f66487b78f58234eca0f88e2adbe80bae4807b809");
+		const INVULNERABLE_E: [u8; 32] =
+			hex!("6c642fb4b571a5685a869cd291fafd575be47a918b231ba28165e5c0cd0cfa15");
+
+		pub struct InitInvulnerables<T>(sp_std::marker::PhantomData<T>);
+		impl<T> OnRuntimeUpgrade for InitInvulnerables<T>
+		where
+			T: frame_system::Config
+				+ pallet_collator_selection::Config
+				+ pallet_session::Config
+				+ pallet_balances::Config,
+			<T as frame_system::Config>::AccountId: From<[u8; 32]>,
+			<T as pallet_session::Config>::ValidatorId: From<[u8; 32]>,
+			<T as pallet_session::Config>::Keys: From<SessionKeys>,
+			<T as pallet_balances::Config>::Balance: From<u128>,
+			<T as pallet_balances::Config>::Balance: EncodeLike<
+				<<T as pallet_collator_selection::Config>::Currency as Currency<
+					<T as frame_system::Config>::AccountId,
+				>>::Balance,
+			>,
+		{
 			#[cfg(feature = "try-runtime")]
 			fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, TryRuntimeError> {
 				Ok(().encode())
 			}
 
 			fn on_runtime_upgrade() -> Weight {
-				let onchain_version = Pallet::<T>::on_chain_storage_version();
-				if onchain_version >= 1 {
-					log::warn!(
-						target: TARGET,
-						"skipping bruteforce to v1 migration: executed on wrong storage version."
-					);
-					return T::DbWeight::get().reads(1)
+				info!(target: TARGET, "initializing the set of invulnerables");
+
+				let raw_keys: Vec<[u8; 32]> = vec![
+					INVULNERABLE_A,
+					INVULNERABLE_B,
+					INVULNERABLE_C,
+					INVULNERABLE_D,
+					INVULNERABLE_E,
+				];
+
+				let validatorids: Vec<<T as pallet_session::Config>::ValidatorId> =
+					raw_keys.iter().map(|&pk| pk.into()).collect();
+
+				pallet_session::Validators::<T>::put(validatorids);
+
+				let queued_keys: Vec<(
+					<T as pallet_session::Config>::ValidatorId,
+					<T as pallet_session::Config>::Keys,
+				)> = raw_keys
+					.iter()
+					.map(|&pk| {
+						(
+							pk.into(),
+							SessionKeys { aura: sr25519::Public::from_raw(pk).into() }.into(),
+						)
+					})
+					.collect();
+
+				pallet_session::QueuedKeys::<T>::put(queued_keys);
+
+				for pk in raw_keys.clone() {
+					pallet_session::NextKeys::<T>::insert::<
+						<T as pallet_session::Config>::ValidatorId,
+						<T as pallet_session::Config>::Keys,
+					>(pk.into(), SessionKeys { aura: sr25519::Public::from_raw(pk).into() }.into());
+					pallet_session::KeyOwner::<T>::insert::<
+						_,
+						<T as pallet_session::Config>::ValidatorId,
+					>((AURA, pk.encode()), pk.into());
```

Comment:
Are the aura keys the same as the validator account IDs?

Reply:
Yes

Reply:
For account ID I mean their on-chain address.

Reply:
according to @georgepisaltu 's comment above, I guess this should work as-is

Reply:
The `ValidatorId` type is actually the same as the `AccountId`, as seen [here](https://github.com/encointer/runtimes/blob/bbf92f696ad8b253ca9f8901ef65022d887fe451/system-parachains/encointer/src/lib.rs#L616).

The aura session key is a wrapper over an `AuraId`, which is defined [here](https://github.com/paritytech/polkadot-sdk/blob/8eb1f5592d15c7c748fbe21c2ea951d3bfe890e2/cumulus/parachains/common/src/lib.rs#L56), being a [`sr25519::Public`](https://github.com/paritytech/polkadot-sdk/blob/8eb1f5592d15c7c748fbe21c2ea951d3bfe890e2/substrate/primitives/consensus/aura/src/lib.rs#L46).

It seems to me everything should match, WDYT @joepetrowski ?

Reply:
Yes the key types match, but my confusion comes from this:

- This block of code is setting the `ValidatorId`s, these are the session keys.
- Just below this, it puts this `BoundedVec` of keys into storage as the `Invulnerable`s. That means you are using the session key for the on-chain account.

For sure it is possible, it's just highly not recommended because you can only change the Invulnerables via governance, but anyone running a collator node should be able to change their session keys. So usually people would have different levels of security on these two keys.

Reply:
You can just ask each of those collators to generate a session key pair and give you the public so you can use them alongside the account public keys.